session recording websocket

ibeckermayer opened the pull request:
Adds a `playbackState` struct which is used to store and manage the playback state over the websocket.

Can be tested against webapps branch [`isaiah/desktop-session-recording`](https://github.com/gravitational/webapps/tree/isaiah/desktop-session-recording) (WIP)

Commit:
Adds the SessionRecording field to WindowsDesktopSessionEnd event to mimic SessionEnd events (useful for easy integration with frontend).

Commit:
14 should have been 12

Commit:
removing test logic

Commit:
Updates the websocket address

Commit:
switches SessionRecording to simple boolean Recorded

Commit:
updates desktopPlaybackHandle to restart playback once it reaches the end

Commit:
adds playback state and synchronization logic for ensuring that goroutines aren't leaked

Commit:
adds toggle functionality for play/pause

ibeckermayer commented on lib/web/desktop.go:
```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+					if evt == nil {
+						h.log.Debug("reached end of playback, restarting")
+						lastDelay = 0
+						// TODO: this causes the session to be re-downloaded, make it smarter by checking for the file first?
```
I'm not sure if this change is easy, but if it is we should definitely do it to avoid re-downloading giant sessions during the normal course of playback.

zmb3 replied:
Restarting when you get to the end seems odd to me. Can we just leave playback at the end and require the page be refreshed if they want to play it again? (at least until we support scrubbing to a specific time)

zmb3 commented on lib/web/desktop.go:
```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+}
+
+// GetPlaying returns ps.playing
+func (ps *playbackState) GetPlaying() bool {
```
`IsPlaying` instead?

zmb3 commented on lib/web/desktop.go:
```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+// in order to ensure that when one goroutine closes, all the others do too
+// (when used in conjunction with GetWsOpen).
+func (ps *playbackState) Close() {
+	ps.ws.Close()
```
Is it okay to close the websocket connection multiple times? Do we see any errors or anything if this happens?

ibeckermayer replied:
If you `Close` a websocket that's already closed it returns a `net.ErrClosed` (this implementation was just ignoring any `Close` errors). I've added error handling and wrapped this with a `sync.Once`.

zmb3 commented on lib/web/desktop.go:
```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+
+				action := playbackAction{}
+				// websocket.JSON.Receive will return even if there's no
+				// message waiting on the websocket, in which case action will
```
Does this mean we're spinning in a tight loop here?

ibeckermayer replied:
No, that comment is totally false -- instead it works more like you'd want it to, and the call hangs until message data is received (or it hits an `io.EOF` error due to websocket being closed). Updated.

zmb3 commented on lib/web/desktop.go:
```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+					return
+				}
+
+				if !ps.GetPlaying() {
```
This will spin and waste CPU when we're paused. Any way we can wait on a channel or condvar when playback is paused instead?

zmb3 commented on lib/web/desktop.go:
```diff
@@ -173,48 +174,175 @@ func proxyWebsocketConn(ws *websocket.Conn, con net.Conn) error {
+		// Hang until the request context is cancelled or the websocket is closed by another goroutine.
+		for {
+			// Return if websocket is closed by another goroutine.
+			if !ps.GetWsOpen() {
```
Same here. Lots of really tight loops that check a boolean as quickly as possible. Let's come up with a proper signaling mechanism to alert the playback routine of the current state.

ibeckermayer commented on api/types/events/events.proto:
```diff
@@ -1699,6 +1699,8 @@ message WindowsDesktopSessionEnd {
     ];
     // DesktopName is the name of the desktop resource.
     string DesktopName = 11 [ (gogoproto.jsontag) = "desktop_name" ];
+    // Recorded is true if the session was recorded, false otherwise.
+    bool Recorded = 12 [ (gogoproto.jsontag) = "recorded" ];
```
Not sure why these are showing up as changes, since [they're already on `zmb3/desktop-session-recording`](https://github.com/gravitational/teleport/blob/zmb3/desktop-session-recording/api/types/events/events.proto#L1702-L1703)

zmb3 replied:
Probably because of a force push to the other branch. I'll stop doing that.

Commit:
fix for the fact that urls are case insensitive

Commit:
moves desktop_playback to its own file, fixes mistaken comment about how websocket.JSON.Receive works, fixes error messaging, wraps playbackState.Close in a sync.Once

Commit:
Adds a cond variable for the two goroutines, but doesn't solve the spinning loop problem for the hanging logic

Commit:
Adds a cancel-able context which is cancelled in ps.Close() in order to avoid a spinning loop in the websocket.Handler

Commit:
Moves the majority of playback logic into the playbackState, which is now renamed to the more accurate playbackPlayer.

zmb3 commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,248 @@
+// hangWhilePaused hangs while pp.playState = playStatePaused. When hangWhilePaused is called in one goroutine,
+// it will only be triggered to check its condition again by another goroutine calling pp.TogglePlaying() or pp.Close()
+// (because those functions call pp.cond.Broadcast()).
+func (pp *playbackPlayer) hangWhilePaused() {
```
Let's call this `waitForPlay` or something similar. Hang has a negative connotation to it.

zmb3 commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,248 @@
+	ws        *websocket.Conn
+	playState string // one of playStatePlaying | playStatePaused | playStateFinished
+	cond      *sync.Cond
+	mu        *sync.RWMutex
```
```suggestion
	mu        sync.RWMutex
```

Generally you don't want a pointer to mutex.

ibeckermayer replied:
The reason this is a pointer is because that mutex is allocated (via a `var mu sync.RWMutex`) first in `NewPlaybackPlayer` before being passed into the `playbackPlayer` struct. If I don't make it a pointer, I get a `go vet` warning on [line 79](https://github.com/gravitational/teleport/blob/isaiah/session-recording-websocket/lib/web/desktop_playback.go#L79) saying `literal copies lock value from mu: sync.RWMutex (copylocks)`

zmb3 commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,248 @@
+				// Only a warning as we expect this case if the websocket is
+				// closed by another goroutine (net.ErrClosed) or by the browser (io.EOF)
+				// while websocket.JSON.Receive() is hanging.
+				pp.log.WithError(err).Warn("error reading from websocket")
```
I don't think it's even a warning then. This is the natural flow when the connection is closed intentionally.

zmb3 commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,248 @@
+	websocket.Handler(func(ws *websocket.Conn) {
+		defer h.log.Debug("desktopPlaybackHandle websocket handler returned")
+
+		pp := NewPlaybackPlayer(sID, ws, ctx.clt, r.Context(), h.log)
```
It might be a little cleaner to not pass the context to new, and instead pass it as the first argument to `ReceiveActions` and `StreamSessionEvents`.

Context should usually be passed as an argument and rarely "stored" in a struct. Plus, in this case, `New` is not the long running operation that needs to support cancelation, `ReceiveActions` and `StreamSessionEvents` are.

(We tend to store context in structs all the time in this code base, so there are exceptions, but I'm trying to get us away from this pattern.)

Commit:
changes pp.hangWhilePaused to pp.waitWhilePaused

Commit:
Moves the context out of NewPlaybackPlayer and the playbackPlayer
struct, wraps playback goroutines in playbackPlayer.Play(ctx) in
order to comply with context semantics.

Commit:
removing unnecessary warnings

zmb3 commented on lib/web/desktop_playback.go:
```diff
@@ -0,0 +1,237 @@
+}
+
+// wait waits until the playbackPlayer's context is cancelled.
+func (pp *playbackPlayer) wait(ctx context.Context, cancel context.CancelFunc) {
```
I would remove the wait func entirely and inline its logic at the end of `Play()`.

Since wait now takes an arbitrary context, there's no guarantee that the caller passes the same context used by the player, making this easy to mis-use.

zmb3 reviewed (approved):
Couple minor comments, but overall this looks great!

Commit:
touchups
